### PR TITLE
export docs and related docs

### PR DESCRIPTION
**Pull request by ETLaurent**
## Summary

- Depends on https://github.com/apostrophecms/apostrophe/pull/4270
- Read PR comments, TODOs and FIXMEs

Big choucroute to batch-export documents, export a single document and page.

Export their actual data that is in the database (we don't use the manager that is populating fields and adding data).
We export the docs as they are in the DB.

We only use the docs manager to get the related docs.

Then we compress the docs in a `aposDocs.json` that's inside a ZIP or GZIP archive, and we return the URL to the client that's opened by the browser for download.

## What are the specific steps to test this change?

single doc (you can also test with a page):

image

image

image

multiple docs:

image

image

image

image

image

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -16,56 +18,17 @@ module.exports = {
...
 
   init(self) {
-    self.apos.asset.iconMap['apos-import-export-download-icon'] = 'Download';
-    self.apos.asset.iconMap['apos-import-export-upload-icon'] = 'Upload';
+    if (self.options.export !== false) {
+      self.apos.asset.iconMap['apos-import-export-download-icon'] = 'Download';
+    }
+    if (self.options.import !== false) {
+      self.apos.asset.iconMap['apos-import-export-upload-icon'] = 'Upload';
+    }
```
Comment (ETLaurent) [inline index.js:26]: not related to this PR, but improves it

File: lib/methods.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      const ids = self.apos.launder.ids(req.query._ids);
+      const relatedTypes = self.apos.launder.strings(req.query.relatedTypes);
+
+      const allIds = self.getAllModesIds(ids);
+
+      if (!relatedTypes.length) {
+        return self.fetchActualDocs(allIds);
+      }
+
+      console.log('==> RELATED TYPED ==>');
```
Comment (ETLaurent) [inline lib/methods.js:17]: let's keep all the logs for now, they will be removed in another PR

File: lib/methods/export.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      const filename = `${self.apos.util.generateId()}-export.${specificExtension || extension}`;
+      const filepath = path.join(self.apos.attachment.uploadfs.getTempPath(), filename);
+
+      const docsData = JSON.stringify(docs, undefined, 2);
+      const attachmentsData = JSON.stringify(attachments, undefined, 2);
+
+      const data = {
+        'aposDocs.json': docsData,
+        'aposAttachments.json': attachmentsData
+        // attachments: 'attachments/' // TODO: add attachment into an "/attachments" folder
```
Comment (ETLaurent) [inline lib/methods/export.js:25]: next step

File: lib/methods/export.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+      const filename = `${self.apos.util.generateId()}-export.${specificExtension || extension}`;
+      const filepath = path.join(self.apos.attachment.uploadfs.getTempPath(), filename);
+
+      const docsData = JSON.stringify(docs, undefined, 2);
+      const attachmentsData = JSON.stringify(attachments, undefined, 2);
+
+      const data = {
+        'aposDocs.json': docsData,
+        'aposAttachments.json': attachmentsData
```
Comment (ETLaurent) [inline lib/methods/export.js:24]: aposAttachment.json has mock data (see TODOs)

File: lib/methods/export.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+      // TODO: use cron instead?
+      // Report is available for one hour by default
+      setTimeout(() => {
+        self.apos.attachment.uploadfs.remove(downloadPath, error => {
+          if (error) {
+            self.apos.util.error(error);
+          }
+        });
+      }, expiration || 1000 * 60 * 60);
```
Comment (ETLaurent) [inline lib/methods/export.js:80]: setTimeout is bad... if the process gets killed, that won't be clearing the downloads

File: lib/methods/export.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+      return {
+        url: downloadUrl
+      };
```
Comment (ETLaurent) [inline lib/methods/export.js:84]: return url, we're not forced to do this because the url is given and opened by the `export-download` event, but let's return something in case of...

File: lib/methods/index.js [hunk 7]
```diff
@@ -0,0 +1,195 @@
+const exportMethods = require('./export');
+
+// TODO: remove:
+const attachmentsMock = [ { foo: 'bar' } ];
```
Comment (ETLaurent) [inline lib/methods/index.js:4]: this TODO

File: lib/methods/index.js [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+      if (reporting) {
+        reporting.setTotal(req.body._ids.length);
+      }
+
+      const ids = self.apos.launder.ids(req.body._ids);
+      const relatedTypes = self.apos.launder.strings(req.body.relatedTypes);
+      const extension = self.apos.launder.string(req.body.extension, 'zip');
+
+      // TODO: add batchSize?
```
Comment (ETLaurent) [inline lib/methods/index.js:21]: ??

File: lib/methods/index.js [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+      if (!self.exportFormats[extension]) {
+        throw self.apos.error('invalid');
+      }
+
+      const allIds = self.getAllModesIds(ids);
+
+      if (!relatedTypes.length) {
+        const docs = await self.fetchActualDocs(req, allIds, reporting);
+
+        // TODO: change undefined to attachments
```
Comment (ETLaurent) [inline lib/methods/index.js:34]: and this TODO

File: lib/methods/index.js [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+      // since they might have different related documents.
+      const relatedIds = draftDocs
+        .concat(publishedDocs)
+        .flatMap(doc => self.getRelatedDocsIds(manager, doc, relatedTypes));
+
+      const allRelatedIds = self.getAllModesIds(relatedIds);
+
+      const docs = await self.fetchActualDocs(req, [ ...allIds, ...allRelatedIds ], reporting);
+
+      // TODO: change undefined to attachments
```
Comment (ETLaurent) [inline lib/methods/index.js:54]: and that TODO

File: lib/methods/index.js [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+      // Use `doc.type` for pages to get the actual schema of the corresponding page type.
+      const schema = manager.schema || self.apos.modules[doc.type].schema;
+
+      return self
+        .getRelatedBySchema(doc, schema)
+        .filter(relatedDoc => relatedTypes.includes(relatedDoc.type))
+        .map(relatedDoc => relatedDoc._id);
+    },
+
+    // TODO: factorize with the one from AposI18nLocalize.vue
```
Comment (ETLaurent) [inline lib/methods/index.js:129]: @ValJed this TODO can be removed when we'll merge our work

File: lib/methods/index.js [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    // TODO: limit recursion to 10 as we do when retrieving related types?
```
Comment (ETLaurent) [inline lib/methods/index.js:130]: ??

File: lib/methods/index.js [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        }
+      }
+      // Filter out doc types that opt out completely (pages should
+      // never be considered "related" to other pages simply because
+      // of navigation links, the feature is meant for pieces that feel more like
+      // part of the document being localized)
+      return related.filter(doc => self.apos.modules[doc.type].relatedDocument !== false);
+    },
+
+    // TODO: use actual attachments when calling this method:
```
Comment (ETLaurent) [inline lib/methods/index.js:171]: and that TODO

File: ui/apos/apps/index.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+export default () => {
+  window.apos.util.onReady(openExportUrl);
+
+  function openExportUrl() {
+    window.apos.bus && window.apos.bus.$on('export-download', event => {
+      if (!event.url) {
+        return;
+      }
+      window.open(event.url, '_blank');
+    });
```
Comment (ETLaurent) [inline ui/apos/apps/index.js:10]: the event I was referring to above

File: lib/methods/index.js [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+        .find({
+          _id: {
+            $in: docsIdsUniq
+          }
+        })
+        .toArray();
+
+      // FIXME: seems like reporting is not working very well,
+      // when calling failure on purpose, we have the progress bar
+      // at 200%... 🤷‍♂️
```
Comment (ETLaurent) [inline lib/methods/index.js:89]: to verify when making final tests

File: index.js [hunk 16]
```diff
@@ -16,56 +20,25 @@ module.exports = {
...
 
-          const relatedTypes = schema
-            .flatMap(searchRelationships)
-            .filter(Boolean);
+    self.exportFormats = {
+      zip,
+      gzip,
+      ...(self.options.exportFormats || {})
```
Comment (ETLaurent) [inline index.js:33]: projects can add custom formats

File: modules/@apostrophecms/import-export-piece-type/index.js [hunk 17]
```diff
@@ -36,7 +36,10 @@ module.exports = {
         export: {
           label: 'aposImportExport:export',
           messages: {
-            progress: 'aposImportExport:exporting'
+            progress: 'aposImportExport:exporting',
+            completed: 'aposImportExport:exported',
+            icon: 'database-export-icon',
```
Comment (ETLaurent) [inline modules/@apostrophecms/import-export-piece-type/index.js:41]: already by default in the core, which is weird and too close to the `piece-type-exporter` module.

still set it here because it's this module responsibility to set that database export icon

File: ui/apos/components/AposExportModal.vue [hunk 18]
```diff
@@ -28,13 +28,10 @@
...
-                disabled
-                :button="{
-                  label: 'ZIP',
-                  icon: 'chevron-down-icon',
-                  modifiers: ['icon-right', 'disabled']
-                }"
+              <AposSelect
+                :choices="extensions"
+                :selected="extension"
+                @change="onExtensionChange"
```
Comment (ETLaurent) [inline ui/apos/components/AposExportModal.vue:34]: yeah, we said only ZIP...
but since gzip was added for the sake of testing another library, let's keep these 2 formats

File: lib/methods.js (hunk 2 above)
Comment (ValJed) [reply lib/methods.js]: Why?

File: lib/methods/export.js (hunk 5 above)
Comment (ValJed) [reply lib/methods/export.js:80]: So we should find a better way to clean files.. is this cleaning files every hours? 
Couldn't we clean the file when the client downloaded it 🤔 ?

File: modules/@apostrophecms/import-export-piece-type/index.js [hunk 19]
```diff
@@ -36,7 +36,10 @@ module.exports = {
...
+            resultsEventName: 'export-download'
```
[the remark this answers is not in the document]
Comment (ValJed) [reply modules/@apostrophecms/import-export-piece-type/index.js:42]: We will need to document properly all these changes before to release.

**Review by ValJed — APPROVED** [review]
